Add extension points to remediate index metadata in during snapshot restore

jbaiera opened the pull request:
Adds a new SPI to allow plugins to hook into the snapshot restore process in order to remediate index metadata before accepting it into the cluster state.

jbaiera committed:
Add extension points to remediate index settings

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

elasticsearchmachine commented:
Hi @jbaiera, I've created a changelog YAML for you.

jbaiera committed:
Update docs/changelog/131706.yaml

Commit:
[CI] Auto commit changes from spotless

DaveCTurner reviewed (changes requested):
Production change seems fine to me. However I don't think `SnapshotResiliencyTests` are quite the right place for this. Instead, could we add a test case to `RestoreSnapshotIT` (or maybe another `AbstractSnapshotIntegTestCase`?) covering this specific change. I'd want us to be injecting the transformer via a plugin (i.e. overriding `org.elasticsearch.test.ESIntegTestCase#nodePlugins`).

DaveCTurner reviewed (commented):
Oh yes also this is a very temporary change right? We can drop it as soon as all the affected snapshots have expired. It'd be worth mentioning this in the javadocs.

dakrone commented on server/src/main/java/org/elasticsearch/snapshots/IndexMetadataRestoreTransformer.java:
```diff
@@ -0,0 +1,23 @@
+
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+
+public interface IndexMetadataRestoreTransformer {
```
Can you add javadocs for this interface?

jbaiera committed:
PR feedback

Add javadoc to new interface, move tests into their own locate outside the resiliency tests.

jbaiera committed:
Cleanup instance creation

jbaiera committed:
Merge branch 'main' into remediate-failure-store-allocation-settings

Commit:
[CI] Auto commit changes from spotless

dakrone reviewed (approved):
LGTM

DaveCTurner reviewed (approved):
LGTM2
